File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

def is_valid_parentheses(text):
    stack = []
    for c in text:
        if c == '(':
            stack.append(c)
        else :
            if not stack:
                return False
            stack.pop()
    return len(stack) == 0

for _ in range(int(input())):
    text = input().strip()
    if is_valid_parentheses(text):
        print("YES")
    else:
        print("NO")

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
using namespace std;


string solve() {
    stack<char> st;
    string input; cin >> input;

    for (char ch : input) {
        if (ch == '(') st.push(ch);
        else {
            if (st.empty()) return "NO";
            st.pop();
        }
    }

    if (!st.empty())  return "NO";
    return "YES";
}
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int T; cin >> T;
    for (int t = 0; t < T; t++) {
        cout << solve() << '\n';
    }

}
